service/agentconf.go

```diff
@@ -0,0 +1,285 @@
+	defer func() {
+		if err != nil {
+			errors.Annotate(err, "failed to copy tools")
+		}
+	}()
+
+	// Setup new and old version.Binarys with only the series
+	// different.
+	fromVers := version.Binary{
+		Number: jujuVersion,
+		Arch:   arch.HostArch(),
+		Series: fromSeries,
+	}
+	toVers := version.Binary{
+		Number: jujuVersion,
+		Arch:   arch.HostArch(),
+		Series: toSeries,
+	}
+
+	// If tools with the new series don't already exist, copy
+	// current tools to new directory with correct series.
+	if _, err = os.Stat(tools.SharedToolsDir(dataDir, toVers)); err != nil {
+		// Copy tools to new directory with correct series.
+		if err = fs.Copy(tools.SharedToolsDir(dataDir, fromVers), tools.SharedToolsDir(dataDir, toVers)); err != nil {
+			return err
+		}
+	}
+
+	// Write tools metadata with new version, however don't change
+	// the URL, so we know where it came from.
+	jujuTools, err := tools.ReadTools(dataDir, toVers)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Only write once
+	if jujuTools.Version != toVers {
+		jujuTools.Version = toVers
+		if err = tools.WriteToolsMetadataData(tools.ToolsDir(dataDir, toVers.String()), jujuTools); err != nil {
+			return err
+		}
+	}
+
+	// Update Agent Tool links
+	var lastError error
+	for _, agentName := range append(unitAgents, machineAgent) {
+		toolPath := tools.ToolsDir(dataDir, toVers.String())
+		toolsDir := tools.ToolsDir(dataDir, agentName)
+
+		err = symlink.Replace(toolsDir, toolPath)
+		if err != nil {
+			lastError = err
+		}
+	}
+
+	return lastError
+}
+
+// StartAllAgents starts all the agents in the machine with specified series.
+func StartAllAgents(machineAgent string, unitAgents []string, dataDir string, series string) (string, []string, error) {
+	var errMachineName string
+	var errUnitNames []string
+
+	running, err := systemd.IsRunning()
+
+	switch {
+	case err != nil:
+		return "", nil, err
+	case !running:
+		return "", nil, errors.Errorf("systemd is not fully running, please reboot to start agents")
+	}
+
+	for _, unit := range unitAgents {
+		if err = startAgent(unit, AgentKindUnit, dataDir, series); err != nil {
+			return "", nil, errors.Annotatef(err, "failed to start %s service", "jujud-"+unit)
+		}
+		errUnitNames = append(errUnitNames, ("jujud-" + unit))
+		logger.Infof("started %s service", "jujud-"+unit)
+	}
+
+	err = startAgent(machineAgent, AgentKindMachine, dataDir, series)
+	if err == nil {
+		errMachineName = "jujud-" + machineAgent
+		logger.Infof("started %s service", "jujud-"+machineAgent)
+	}
+	return errMachineName, errUnitNames, errors.Annotatef(err, "failed to start %s service", "jujud-"+machineAgent)
+}
+
+func startAgent(name string, kind AgentKind, dataDir string, series string) (err error) {
+	renderer, err := shell.NewRenderer("")
+	if err != nil {
+		return err
+	}
+	info := NewAgentInfo(
+		kind,
+		name,
+		dataDir,
+		paths.MustSucceed(paths.LogDir(series)),
+	)
+	conf := AgentConf(info, renderer)
+	svcName := "jujud-" + name
```

Comment:
There's several places where we go "jujud-" + name
There could already be a helper in the code, but if not use a helper for this, eg
func serviceName(agent string) string {
    return "jujud-" + agent
}